### DISCUSSION TITLE
list issued documents aziende clienti

### DISCUSSION THREAD

**Discussion by programmer1855** [fattureincloud/api#172] [Q&A] [answered]
Buonasera,
devo recuperare le fatture delle aziende clienti, ho creato l'applicazione come mostrato in immagine, ma quando eseguo la chiamata [List Issued Documents](https://developers.fattureincloud.it/api-reference/#tag--Issued-Documents) riesco ad ottenere solo le fatture dell'azienda principale.
Se nella chiamata metto l'id di una delle aziende clienti, mi restituisce l'errore:
`{
  "error": {
    "message": "Unauthorized."
  }
}`

Specifico che nella generazione del token ho dato tutti i permessi di sola lettura

Il problema è dovuto al fatto che nell'applicazione non sono state aggiunte le aziende clienti? Se sì, come faccio ad aggiungerle dal momento che è possibile selezionare solo l'azienda principale?

![Cattura](https://user-images.githubusercontent.com/119016241/203871564-a1f21963-b717-4568-9603-c9e03162e7f2.PNG)
![screen](https://user-images.githubusercontent.com/119016241/203871566-f6c9ae76-770d-46db-b0bf-1a307e3d29df.PNG)

**Comment by emajo** [comment]
Ciao @programmer1855, attualmente non sono esposte chiamate per i commercialisti, si potrebbe raggiungere un risultato simile creando un applicazione a cui far collegare tutti gli attuali clienti, per andare a recuperare un token per ognuno di loro e utilizzarlo per recuperare le fatture.

Grazie,
Emanuele.

**Comment by programmer1855** [comment, marked as answer]
Buongiorno Emanuele,
ho provato a fare come proposto ma non funziona.

Ho creato un'applicazione dall'azienda principale e mi sono salvato il suo id applicazione, ma quando entro nel profilo dell'azienda cliente e collego l'app, mi mostra la stessa schermata che ho allegato a questo topic, una finestra modale con solo l'azienda principale...
Per noi è indispensabile far collegare le aziende clienti ad almeno una applicazione con cui faremo delle chiamate per recuperare il relativo fatturato, è possibile parlare con un vostro tecnico?

Grazie

**Reply by valmoz** [reply]
Buongiorno,
io e @emajo siamo parte del team API di Fatture in Cloud, quindi siamo le persone più qualificate per risponderle.

Il processo che deve seguire è il seguente:
1. Se la sua applicazione è privata (come penso), deve aggiungere la mail dell'account dei suoi clienti nella whitelist dell'applicazione (per maggiori informazioni: [Create an app](https://developers.fattureincloud.it/docs/basics/create-an-app#-app-visibility))
2. Deve inviare le seguenti informazioni ai suoi clienti: Client ID dell'applicazione, lista degli [scopes](https://developers.fattureincloud.it/docs/basics/scopes) da associare all'applicazione
3. I suoi clienti, dal loro profilo FIC, devono accedere alla sezione Applicazioni Collegate, inserire il client ID e selezionare gli scope come indicato qui: [Manual Token](https://developers.fattureincloud.it/docs/authentication/manual-authentication)
4. I sui clienti devono copiare l'Access Token generato e passarglielo

Utilizzando tale access token sarà possibile accedere alle API

Mi faccia sapere se si verificano ulteriori problemi.

Grazie
Mauro

**Reply by programmer1855** [reply]
Buongiorno, 
confermo che con questa procedura riesco a recuperare correttamente le fatture delle aziende clienti.

Purtroppo però è impraticabile per le nostre centianaia di aziende clienti, posso suggerire in una prossima versione delle API di automatizzare questo processo.

Ringrazio entrambi per l'aiuto

**Reply by valmoz** [reply]
Lo svantaggio del flusso manual token è proprio quello della macchinosità della procedura.
Abbiamo però altri metodi di autenticazione che può utilizzare per automatizzare il processo di connessione, rendendolo più semplice ai suoi utenti.

Ad esempio, può utilizzare il flusso OAuth2.0 Authentication Code che è pensato esattamente per gestire molteplici account sconosciuti a priori. L'utilizzo di questo flusso è più complicato in quanto richiede degli step aggiuntivi in fase di sviluppo, ma permette di rendere più semplice la creazione e raccolta dei token.

Le allego delle pagine che potrebbero esserle utili:

- [Diagramma per la scelta del metodo di autenticazione](https://developers.fattureincloud.it/docs/basics/authentication)
- [OAuth 2.0 Authentication Code Flow](https://developers.fattureincloud.it/docs/authentication/code-flow/)
- [Manual Token](https://developers.fattureincloud.it/docs/authentication/manual-authentication)
- [OAuth 2.0 Device Code Flow](https://developers.fattureincloud.it/docs/authentication/device-code) (non è generally available)

**Reply by valmoz** [reply]
Buongiorno @programmer1855 le chiederei di farci sapere se uno degli altri meccanismi di autenticazione può fare al caso suo.

**Comment by programmer1855** [comment]
Buonasera,
ho risolto usando il suggerimento di @emajo facendo collegare le aziende clienti all'app, generando un token per ognuna di loro, e facendo chiamate API distinte.

L'unico grosso problema nell'implementare questa procedura è stato che alla generazione dell'access token non ho nessuna informazione di ritorno dell'azienda che lo ha generato, basterebbe anche solo il company_id di fatture in cloud; in questo modo sarebbe possibile capire che quella particolare azienda cliente ha generato quel determinato access token.

Grazie per la collaborazione

**Reply by valmoz** [reply]
Buongiorno @programmer1855 
la chiamata [ListUserCompanies](https://developers.fattureincloud.it/api-reference#get-/user/companies), effettuata con il token, non le è d'aiuto?

Può leggere anche: [Where can I find my company ID?](https://developers.fattureincloud.it/docs/basics/company-scoped-methods#-where-can-i-find-my-company_id)

Grazie
Mauro
